### ISSUE TITLE
401 error on post with special characters

### ISSUE THREAD

**Issue by YacineMessaadi** [Secreto31126/whatsapp-api-js#290]
**Describe the bug**
When I send a message to the API containing special characters, for example "J'ai mangé des pâtes", the post request responds a 401 error.

**Comment by Secreto31126** [comment]
I can already tell this is an issue with the signature verification.

2 questions:

1. Are you sure it's related to the special characters? Does it work as expected if you send a message in plain English?

2. Would you mind sending a snippet of your code to see how you retrieve the raw body from the request? I want to know if there's something that might be breaking the text encoding

In the meantime I will be investigating about signature verification with special characters, maybe it's not directly supported from the JS side or Facebook has some weird rules to handle them.

Thanks for bringing up this bug!

**Comment by Secreto31126** [comment]
> Please note that we generate the signature using an escaped unicode version of the payload, with lowercase hex digits. If you just calculate against the decoded bytes, you will end up with a different signature. For example, the string äöå should be escaped to \u00e4\u00f6\u00e5

https://developers.facebook.com/docs/messenger-platform/webhooks#validate-payloads

It seems like the issue goes this way, I need to find out how to convert the special chars into the escaped unicodes.

**Comment by YacineMessaadi** [comment]
> I can already tell this is an issue with the signature verification.
> 
> 2 questions:
> 
> 1. Are you sure it's related to the special characters? Does it work as expected if you send a message in plain English?
> 2. Would you mind sending a snippet of your code to see how you retrieve the raw body from the request? I want to know if there's something that might be breaking the text encoding
> 
> In the meantime I will be investigating about signature verification with special characters, maybe it's not directly supported from the JS side or Facebook has some weird rules to handle them.
> 
> Thanks for bringing up this bug!

Thanks for your fast reply !

I am sure that it's related to special characters because plain english messages work as expected !

There is a snippet of my code (in Typescript Nestjs):
`@Post('webhook')
  async post(@Req() request: Request): Promise<string> {
    let data = JSON.stringify(request.body);

    const status = await this.appService.post({
      data: data,
      headers: request.headers as any,
    });

    if (status === 200) {
      return 'OK';
    }
    return 'ERROR';
  }`

**Comment by Secreto31126** [comment]
Ok, so it definitely seems like an issue from the library side.

I already stole a code snippet from StackOverflow which should fix the issue, in a few minutes I will have a PR ready to merge.

Side note: Using JSON.stringify to get the raw body is prone to breaking, as it might format the output slightly different as the origial, and thus throwing 401 because the signature will not match. AFAIK, in NestJS there's no alternative unfortunately. Just mentioning it for future reference, in case it ever brings you issues.

**Comment by Secreto31126** [comment]
So, about those few minutes... I got a little sidetracked :)

Tomorrow after I merge #289 I will release 2.5.0 which will include the fix for the bug. When you get to try it, let me know if it works as expected.

Thanks!
